### PR TITLE
refactor(charts): fixed charts responsive with refeactoring

### PR DESCRIPTION
**Pull request by neighborhood999**
### Single Report
https://user-images.githubusercontent.com/10325111/176365110-416f4f71-0784-48bd-97ae-9b5e2fea538c.mp4

### Comparison Report
https://user-images.githubusercontent.com/10325111/176365141-ca7bfbe7-1e9d-4e8e-82d3-9e63d077aa3e.mp4

## Description

- Fixed charts responsive issues
- Refactored messy code
- Added hooks for rendering single & comparison report
- Adjustment scripts for running single/comparison at the same time with different ports

Signed-off-by: Jie Peng <im@jiepeng.me>

### CODE REVIEW

File: static_report/src/hooks/useComparisonChart.ts [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    const xSubGroup = d3
+      .scaleBand()
+      .domain(GROUPED)
+      .range([0, xScale.bandwidth()])
+      .padding(X_SUB_PADDING);
+
+    const xAxis = d3.axisBottom(xScale).tickFormat((value, i) => {
+      const xAxisItemLength = xScale.domain().length - 1;
+
+      if (i === 0 || i === xAxisItemLength / 2 || i === xAxisItemLength) {
```
Comment (jonycfu) [inline static_report/src/hooks/useComparisonChart.ts:46]: For the middle ticker, should we handle both odd/even ticker lengths?
image

File: static_report/src/utils/index.tsx [hunk 2]
```diff
@@ -134,3 +136,113 @@ export function extractExpectedOrActual(value) {
 
   return value;
 }
+
+// for comparison
+export function joinBykey(base = {}, input = {}) {
+  const result = {};
+
+  Object.entries(base).forEach(([key, value]) => {
```
Comment (jonycfu) [inline static_report/src/utils/index.tsx:144]: these two `forEach(Fn)` operations can be refactored.

File: static_report/src/utils/index.tsx [hunk 3]
```diff
@@ -134,3 +136,113 @@ export function extractExpectedOrActual(value) {
 
   return value;
 }
+
+// for comparison
+export function joinBykey(base = {}, input = {}) {
```
Comment (jonycfu) [inline static_report/src/utils/index.tsx:141]: This name is a little confusing. Maybe rename as `nestComparisonValueByKey` ? (not the best, but I at least have more context)
Or at least explain more in the comment, as it's not obvious

File: static_report/src/utils/index.tsx [hunk 4]
```diff
@@ -134,3 +136,113 @@ export function extractExpectedOrActual(value) {
...
+    if (!result[key]) {
+      result[key] = {};
+    }
+    result[key]['input'] = value;
+  });
+
+  return result;
+}
+
+export function getComparisonTests(assertion, from) {
```
Comment (jonycfu) [inline static_report/src/utils/index.tsx:162]: `getComparisonAssertionTests`

File: static_report/src/utils/index.tsx [hunk 5]
```diff
@@ -134,3 +136,113 @@ export function extractExpectedOrActual(value) {
...
+  const { passed, failed } = getReportAsserationStatusCounts(assertion);
+
+  if (assertion) {
```
Comment (jonycfu) [inline static_report/src/utils/index.tsx:165]: I think checking for `!assertion` (base case) up top first would make this more readable

**Review by jonycfu — COMMENTED** [review]
minor tidy up.
I do think the `utility/index.tsx` file needs functional typing. Because, otherwise, it's hard for future-me future-you to know how to reuse certain methods

File: static_report/src/hooks/useComparisonChart.ts (hunk 1 above)
Comment (neighborhood999) [reply static_report/src/hooks/useComparisonChart.ts:46]: Clarify with Jon, resolved.

File: static_report/src/utils/index.tsx [hunk 6]
```diff
@@ -138,59 +145,77 @@ export function extractExpectedOrActual(value) {
...
+  { base: Record<string, unknown>; input: Record<string, unknown> }
+> {
+  const _base = Object.entries(base).reduce((acc, [key, value]) => {
+    if (!acc[key]) {
+      acc[key] = Object.create({});
     }
 
-    result[key]['base'] = value;
-  });
+    acc[key] = value;
```
Comment (jonycfu) [inline static_report/src/utils/index.tsx:160]: hmm, did we lose the `['base' | 'input']` key access here?

Comment (jonycfu) [reply static_report/src/utils/index.tsx:160]: perhaps write a function for `.reduce(myFunc('input'), {})`
```
const myFunc = (type) => (acc, [key, val]) => { 
  acc[key] = acc[key][type] ? Object.create({}) : value } 
  return acc;
}
```

**Review by jonycfu — CHANGES_REQUESTED** [review]
leftovers

File: static_report/src/utils/index.tsx (hunk 6 above)
Comment (neighborhood999) [reply static_report/src/utils/index.tsx:160]: In the `mergeWith`:

```js
{
  base: base,
  input: input
}
```

**Review by jonycfu — APPROVED** [review]

**Pull request merged**